### PR TITLE
math: Add a simple fixed/floating point math library

### PR DESCRIPTION
**Pull request by yperess**
Add a fixed/floating point math library via a header that can be configured via Kconfigs.

### CODE REVIEW

File: Kconfig.headers
Comment (teburd) [inline Kconfig.headers:8]: I'm pretty skeptical of this Kconfig option after having seen the results where the math may work only in one case and not the other.

File: Kconfig.headers
Comment (teburd) [inline Kconfig.headers:15]: Similar to the above comment really, I don't believe a Kconfig option should be drastically changing the way in which project wide math works.

If varying q formats are useful, then perhaps make those as new types. If you perhaps assume _Generic exists it makes it maybe even easier to write a common fp_div/fp_mul and perhaps some saturating operations which are useful. These could wrap format specific versions.

I don't really think we need to enable every fixed point format here. A few useful ones would be fine. e.g. sensor_value could be replaced by typedef uint64_t q31_32_t easily and effectively fit all the known sensors in the tree. For normalized values q1.15, q1.31 are likely *plenty* as shown by arm's cmsis-dsp.

e.g. typedef uint32_t q15_16_t;

So maybe again, a short list of fixed point formats, matching format strings, mul/div, and conversion functions to float, integers, and maybe integer with a divider.

An example maybe helps clarify my thinking here...

I have an int representing 3 microsecond, I want to convert it to fixed point with seconds as the measurement unit.

Ideally any conversion should validate the value is at least somewhat representable.

So perhaps mn_to_q15_16(3, 1000000); should fail on an assert.

Something like mn_to_q31_32(3, 1000000); or mn_to_q1_31(3, 1000000); should be ok.

What I don't want is this which bit me once already...

fp_t seconds = fp_div(INT_TO_FP(3), INT_TO_FP(1000000));

What should not happen, in my opinion, is the above sort of conversion failing or succeeding based on a project wide Kconfig.

**Review by teburd — CHANGES_REQUESTED** [review]

**Comment by teburd** [comment]
I want to be clear that I think having some fixed point types and math defined in the tree would be really helpful. There are places that fixed point math could resolve precision issues with integers, and avoid custom types for interfaces (sensor_value). I just think its important the behavior be non-surprising depending on Kconfig options if that makes sense

File: Kconfig.headers
Comment (yperess) [reply Kconfig.headers:15]: I can easily do this in C++ using a template but I can't think of a good way to do this in C.  Open to suggestions. I worry about typedef as it can still be mixed and cause errors.

File: include/zephyr/math/fp.h
Comment (pdgendt) [inline include/zephyr/math/fp.h:88]: I think the function definition has to stay the same regardless of `Kconfig` options. Avoid code duplication and have a single definition.

```suggestion
static inline fp_t fp_mul(fp_t a, fp_t b)
{
#ifdef CONFIG_MATH_UTIL_FLOAT
	return a * b;
#else
	return (fp_t)(((fp_inter_t)a * b) >> CONFIG_MATH_UTIL_FP_BITS);
#endif
}
```

Comment (yperess) [reply include/zephyr/math/fp.h:88]: Thanks, I'll update today

File: Kconfig.headers
Comment (yperess) [reply Kconfig.headers:15]: Yeah, I'm not finding a good way to do this in C, in C++ it's easy:

```
constexpr fp_t FLOAT_TO_FP(float x) {
  fp_inter_t intermediate = (fp_inter_t)(x * (float)(1 << 16));
  assert(intermediate < FLT_MAX);
  assert(intermediate > FLT_MIN);
  return (fp_t)intermediate;
}
```

Comment (yperess) [reply Kconfig.headers:15]: For making sure that different values aren't combined with different resolution:

```
template <int decimal_bits>
struct fp {
  union {
    int32_t raw;
    struct {
      int32_t integer : 32 - decimal_bits;
      int32_t decimal : decimal_bits;
    };
  };
};
```

Now you can do:
```
typedef struct fp<16> q15_16;
typedef struct fp<8> q23_8;
...
```

Comment (teburd) [reply Kconfig.headers:15]: You can do the same sort of thing with a macro in C, its not as good as a const expr, but a statement expr macro would be close enough


``` c
/** Define a q1_31_t */
typedef struct { int32_t q1_31; } q1_31_t;

/**
 * Convert a value of a numerical rval to a qval
 * and validate it fits in the bounds of the qval
 * close equivalent to q = val; with bounds checking.
 */
#define q_convert(qval, rval) \
do { \
  int32_t whole = z_q_whole(qval, rval); \
  int32_t fraction = z_q_fract(qval, rval); \
  z_q_assign(qval, whole, fract); \
} while (false)
   
```
 
So something like the above perhaps might work out, where you *know* the type of q being assigned and the type you are converting from.

It requires use of _Generic((qval), q1_31_t: ..., ) style macros which... should be fine.

It avoids operator mismatching, everything must be done with macros

q_add, q_sub, q_mul, q_div, q_mod, q_rem, q_abs, etc, etc, etc

And many of them likely require type specialization which can be done with _Generic.

The fewer q types you have, the saner this ends up.

I know its not as sane as say templates, but its possible I believe.

Comment (yperess) [reply Kconfig.headers:15]: `_Generic` is neat, I haven't seen it before. I'll see if I can get this to work.

Comment (pdgendt) [reply Kconfig.headers:15]: Note that `_Generic` is a C11 feature.

Comment (teburd) [reply Kconfig.headers:15]: It wouldn't be the first usage of it in the tree, though it would push us to say _Generic is whats required for this. I don't think that's necessarily a bad thing.

Comment (teburd) [reply Kconfig.headers:15]: Maybe one more nifty feature of this... and probably should be a requirement.

If the hardware supports fixed point instructions of certain q types and operations, these could should easily be overridden with those.

Comment (yperess) [reply Kconfig.headers:15]: C11 is already documented as "highly recommended" https://docs.zephyrproject.org/latest/develop/languages/c/index.html?highlight=c11#c-language-support

File: include/zephyr/math/q_ops.h
Comment (yperess) [inline include/zephyr/math/q_ops.h:33]: @teburd I can't get this part to work. Do you have any ideas? The issue is that `_Generic` needs to compile for every branch of the statements and I can't cast `q1_8_t` to `int64_t`. Alternatively, if I were to use `x.value` this will fail if I pass `INT32_C(5)` for `x`.

Comment (yperess) [reply include/zephyr/math/q_ops.h]: Got it working by using a function pointer

**Review by teburd — COMMENTED** [review]
I think this is showing some serious promise, though do take a look at my notes on the q types and notations.

I think we want to match up closely with the arm types as an initial step which are q7_t, q15_t, and q31_t as mentioned before. But the types here aren't quite the same.

**Comment by yperess** [comment]
@teburd I pushed my latest. Could you take a look and see if this looks like what you had in mind? Keep in mind that tests are still missing for (tests take a very long time to write):
1. `q_div()`
2. All operations using `q1_15_16_t` and `q1_0_31` as the lhs

Missing functionality:
1. C++
2. Printing the fixed point using `PRIf` and `PRIf_arg()`

File: include/zephyr/math/q_ops.h
Comment (teburd) [inline include/zephyr/math/q_ops.h:125]: You can use gcc macro expressions to provide returns instead of mutating the given variable, there's really no other way to do that otherwise that I'm aware of (I don't know everything). If people are using non-gcc compatible (gcc/clang) then perhaps this subsys won't be built or allowed, and that's ok I think.

**Review by teburd — COMMENTED** [review]
This is definitely looking better. The qS_M_N notation might be a bit much, are we planning on having unsigned Q types? Even if we are, the sign bit can be implied

q15_16_t implies a sign bit
q0_31_t implies a sign bit

That's fine to annotate the types like that I'd think, and makes it a little shorter.

I'd avoid mutating the left hand parameter though and use macro expressions to do the math and return a new q value.

File: include/zephyr/math/q_ops.h
Comment (yperess) [reply include/zephyr/math/q_ops.h:122]: Just saw this, I'll make the change to update to return a value instead of mutating tomorrow.

**Comment by yperess** [comment]
@teburd PTAL I got both a C and C++ implementation now (though I'll need to update the type names and make the functions return a `q` value instead of updating the LHS. Also, will need to document all of this.

**Review by MaureenHelm — COMMENTED** [review]
This seems headed toward reinventing CMSIS-DSP. Perhaps instead we should investigate extending what we already have for ARM to additional architectures.

File: include/zephyr/math/q_ops.h
Comment (uLipe) [inline include/zephyr/math/q_ops.h:129]: Is there any reason to use false and depend of stdbool instead of just use while(0) ?

**Review by uLipe — COMMENTED** [review]
One minor comment about the implementation.

Although I like the idea of having a fixed point math module like this one, I went through the CMSIS-DSP and all the operations above are covered there, with the advantage of setting the block size as a parameter, that is it, suitable to scalar or vector data, except for `q_assign` .

File: include/zephyr/math/q_types.h
Comment (yperess) [reply include/zephyr/math/q_types.h:58]: There was an offline discussion. I would have added the details here, I'm sorry.

Basically, Tom was saying that:
- `q1_7_t` is an 8 bit value with 1 sign bit and 7 decimal bits,
- `q1_15_t` is a 16 bit value with 1 sign bit and 15 decimal bits.

For my first pass here I actually wanted all 32 bit values (I was using a short hand to specify it). So:
- When I used `q1_8_t` I meant signed, with 8 bit decimal (assumed 23 bit integer)
- When I used `q1_16_t` I meant signed, with 16 bit decimal (assumed 15 bit integer)

For this PR, I'll be implementing the following splits:
- 1/0/31
- 1/15/16
- 1/23/8

File: include/zephyr/math/q_types.h
Comment (yperess) [reply include/zephyr/math/q_types.h:69]: Done

File: include/zephyr/math/q_types.h
Comment (yperess) [reply include/zephyr/math/q_types.h:33]: Done

File: include/zephyr/math/q_ops.h
Comment (yperess) [reply include/zephyr/math/q_ops.h:129]: Just the coding style

**Comment by yperess** [comment]
@MaureenHelm @uLipe (I'm merging the two comment threads since they both relate to CMSIS-DSP).

I looked at the DSP implementation and there are several things about it that I did not like (which is why I wanted to start from scratch here).
1. This is all done in headers and is better optimized
2. This is strongly typed, you'll get a warning if you try:
```
q1_23_8_t q0 = q_static(q1_23_8_t, 0.5);
q1_0_31_t q1 = q0;
```
3. I need to finish adding the Kconfig for this, but soon I'll update the PR with CONFIG_MATH_FP_ASSERT_ON_OVERFLOW which will allow you to toggle the assert in the macros to make sure your numbers don't overflow if you want.
4. CMSIS provides arithmetic functions add/sub/mul/div per type, it cannot handle `q_add(q8_t, q16_t)` whereas this implementation does without having to declare a new function signature/name for each combination (since C doesn't allow us to overload).

I'm not sure what the best steps forward are with CMSIS, I do see the overlap. If you don't feel that the above are good use cases, I'm happy to abandon this and move the CMSIS library over to Zephyr proper (less the ARM specific features).

Adding @keith-zephyr and @gwendalcr to the discussion

File: include/zephyr/math/q_ops.h
Comment (yperess) [reply include/zephyr/math/q_ops.h:122]: I'm going to hold off doing more work on this until the discussion over CMSIS-DSP is resolved.

**Review by stephanosio — COMMENTED** [review]
> 2\. This is strongly typed, you'll get a warning if you try:

Sure, but it seems to come with its own drawbacks and ugliness; e.g.: `((q1_23_8_t){.value = 0x78})`, `qval.value = intermediate & 0xffffffff;`

Also the `.value` in itself is not "strongly typed," and if you are going to/have to use `.value` everywhere to access its value and do something with it, then this "strongly typed" advantage kind of goes away.

> 4\. CMSIS provides arithmetic functions add/sub/mul/div per type, it cannot handle `q_add(q8_t, q16_t)` whereas this implementation does without having to declare a new function signature/name for each combination (since C doesn't allow us to overload).

Normally, if you are doing binary operations on two different types, you would want to do the proper conversion(s) first such that they are of the same type.

> I'm not sure what the best steps forward are with CMSIS, I do see the overlap. If you don't feel that the above are good use cases, I'm happy to abandon this and move the CMSIS library over to Zephyr proper (less the ARM specific features).

Practically, this is only going to be useful if there is an extensive set of functions/libraries on top of this (e.g. complex numbers, matrices, trigonometry, statistics, ...), which either we (Zephyr) or someone using it will have to implement.

At that point, I agree that it is akin to reinventing CMSIS-DSP and may not be something we want to do and keep maintaining as part of Zephyr.

CMSIS-DSP can and currently does support other architectures because it provides the full C implementation (without any arch-specific intrinsics) of all the functions it supports; for instance, you can build CMSIS-DSP for the host operating system (`native_posix`) for testing purposes -- of course, without any advanced architecture-specific optimisations and SIMD support, it will be painfully slow; but, my point is that CMSIS-DSP in itself is not directly dependent on ARM.

IMHO, my suggestion would be to improve what already exists (i.e. CMSIS-DSP) than to reinvent something of our own. Also, CMSIS-DSP is an open source project licensed under Apache 2.0 after all (by the way, [they now made it an independent component, so it is no longer tied to the "core CMSIS"](https://github.com/ARM-software/CMSIS-DSP)).

**Comment by yperess** [comment]
> > 2. This is strongly typed, you'll get a warning if you try:
> 
> Sure, but it seems to come with its own drawbacks and ugliness; e.g.: `((q1_23_8_t){.value = 0x78})`, `qval.value = intermediate & 0xffffffff;`
> 
> Also the `.value` in itself is not "strongly typed," and if you are going to/have to use `.value` everywhere to access its value and do something with it, then this "strongly typed" advantage kind of goes away.
> 
> > 4. CMSIS provides arithmetic functions add/sub/mul/div per type, it cannot handle `q_add(q8_t, q16_t)` whereas this implementation does without having to declare a new function signature/name for each combination (since C doesn't allow us to overload).
> 
> Normally, if you are doing binary operations on two different types, you would want to do the proper conversion(s) first such that they are of the same type.
> 
> > I'm not sure what the best steps forward are with CMSIS, I do see the overlap. If you don't feel that the above are good use cases, I'm happy to abandon this and move the CMSIS library over to Zephyr proper (less the ARM specific features).
> 
> Practically, this is only going to be useful if there is an extensive set of functions/libraries on top of this (e.g. complex numbers, matrices, trigonometry, statistics, ...), which either we (Zephyr) or someone using it will have to implement.
> 
> At that point, I agree that it is akin to reinventing CMSIS-DSP and may not be something we want to do and keep maintaining as part of Zephyr.
> 
> CMSIS-DSP can and currently does support other architectures because it provides the full C implementation (without any arch-specific intrinsics) of all the functions it supports; for instance, you can build CMSIS-DSP for the host operating system (`native_posix`) for testing purposes -- of course, without any advanced architecture-specific optimisations and SIMD support, it will be painfully slow; but, my point is that CMSIS-DSP in itself is not directly dependent on ARM.
> 
> IMHO, my suggestion would be to improve what already exists (i.e. CMSIS-DSP) than to reinvent something of our own. Also, CMSIS-DSP is an open source project licensed under Apache 2.0 after all (by the way, [they now made it an independent component, so it is no longer tied to the "core CMSIS"](https://github.com/ARM-software/CMSIS-DSP)).

What would be the recommended approach for this? I imagine we'd need to start off by copying the existing `DSP/include` directory into the Zephyr repo and rename the headers to be architecture agnostic. We can then start adding architecture specific optimizations (starting wth arm since those exist already). Does that sound right?

**Comment by stephanosio** [comment]
> What would be the recommended approach for this? I imagine we'd need to start off by copying the existing `DSP/include` directory into the Zephyr repo and rename the headers to be architecture agnostic. We can then start adding architecture specific optimizations (starting wth arm since those exist already). Does that sound right?

Personally, I would at least evaluate the feasibility of working with the CMSIS-DSP maintainers to rename the `arm_` prefix to something more architecture agnostic (e.g. `cmsis_` or `cmsis_dsp_`) so that the changes in our fork will stay more or less incremental, and keep the `arm_` prefix even for other architectures, although not ideal, in the meantime (other than the questionable naming, this does not do any harm to us).

If we rename all the CMSIS-DSP files and functions now, syncing to the upstream and pulling any upstream bug fixes and enhancements will be a complete nightmare. Of course, if the upstream CMSIS-DSP maintainers have no intention of cooperating and/or our fork has progressed to the point that there exist irreconcilable differences and it no longer makes sense to keep ties to the upstream, then it will be a different story.

So my suggestion is to:

1. make incremental changes to the [Zephyr CMSIS-DSP module/fork](https://github.com/zephyrproject-rtos/cmsis) (e.g. adding code for other architecture support)
1. upstream changes to CMSIS-DSP whenever it makes sense

**Comment by cfriedt** [comment]
A `Q<m,n>` class is pretty easy to do in C++ and would be way easier to maintain. There's a lot of opportunity for the compiler to optimize as well.

Could export a C API..

It's just unfortunate that it would be C++ and some have reservations about that.

Otherwise, I think @teburd 's suggestion is probably smart. I.e. use `q7_t` and friends, maybe provide some macros.

**Pull request closed without merging**
